contracts/modules/cover/Cover.sol

```diff
@@ -351,22 +380,44 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon {
 
     require(premiumInPaymentAsset <= buyCoverParams.maxPremiumInAsset, "Cover: Price exceeds maxPremiumInAsset");
 
+    // calculate refundValue in NXM
     uint refundInNXM = refundInCoverAsset * 1e18 / pool().getTokenPrice(cover.payoutAsset);
+    handlePaymentAndRefund(buyCoverParams, totalPremiumInNXM, premiumInPaymentAsset, refundInNXM);
 
-    if (buyCoverParams.payWithNXM) {
-      if (refundInNXM < totalPremiumInNXM) {
-        // requires NXM allowance
-        retrieveNXMPayment(totalPremiumInNXM - refundInNXM, buyCoverParams.commissionRatio, buyCoverParams.commissionDestination);
-      }
-    } else {
-      uint refundInPaymentAsset =
-      refundInNXM
-      * (pool().getTokenPrice(buyCoverParams.payoutAsset) / 10 ** paymentAssetDecimals);
+    updateGlobalActiveCoverAmountPerAsset(buyCoverParams.period, buyCoverParams.amount, buyCoverParams.payoutAsset);
+  }
 
-      if (refundInPaymentAsset < premiumInPaymentAsset) {
-        // retrieve extra required payment
-        retrievePayment(premiumInPaymentAsset - refundInPaymentAsset, buyCoverParams);
-      }
+  function handlePaymentAndRefund(
+    BuyCoverParams memory buyCoverParams,
+    uint totalPremiumInNXM,
+    uint premiumInPaymentAsset,
+    uint refundInNXM
+  ) internal {
+
```

Comment:
you can add an early check for `refundInNXM < totalPremiumInNXM` and exit the function early without performing anything else if we don't need payment.